### ISSUE TITLE
Refactor command line interface

### ISSUE THREAD

**Issue by willu47** [SALib/SALib#216]
## Problem

The command line commands used to run a sample or analysis function involves a lot of code duplication.  Each module contains an `if __name__ == '__main__'` section containing an argument parser, often with duplicate arguments.

## Recommendation

Write one central command-line interface argument parser which interrogates the available methods and exposes the arguments on the command line.  An an entry point for an `salib` command.

For example
```bash
$ salib sample morris

SALib: Sensitivity Analysis Library

Morris Method requires the following arguments
-n <int> number of trajectories
-o <str> filepath to write sample
...

$ salib sample morris -n 10 -o my_sample.csv

```

**Comment by willu47** [comment]
Hi - thanks for this.  I've had a play around with your script.  A few comments:

- adding an entry point from `setup.cfg` is straight forward and provides a neat facility on the command line - great!
- The CLI could do with an up-front help message - check to see if no arguments are passed (e.g. sample/analyze) and display help if not
- The fact that the CLI has expectations for the methods, implies that we need to define an interface for them - perhaps a lightweight abstract class (Sample and Analyze) which each of the methods can inherit and implement.  This may influence the organisation of the library and bring the utility modules (e.g. `common_args`, `util/*`) into a more understandable structure.

This refactoring work needs to take place with the understanding that users also work programmatically with an individual method as well, so the ability to import a module and call `sample` on it, needs to be retained.

Let me know what you think.

**Comment by ConnectedSystems** [comment]
Yes - agree on all counts except maybe the last. Stressing the maybe.

I'm trying to think of an approach that will achieve the goal of improved maintainability without a major restructure around classes. But I understand the appeal of having a common interface that all methods adhere to and it is looking like the cleanest path at the moment.

I just got around to handling the case where no arguments are passed; it now displays the help message. Will push the change up once I've cleaned up a few things.

**Comment by ConnectedSystems** [comment]
To give an update on this - what I'm thinking now is that the current implementation is quite good and a restructure around classes is probably unnecessary. Unless some new sampling/analysis (S/A) method requires it, I'd suggest holding off on refactoring things (which will undoubtedly break some tests as well). 

Inconsistency between S/A methods can occur, granted, but they can occur even with an interface class. Ensuring everything has the right functions/methods can be done through policy and backed by tests. e.g. contributed S/A methods don't get accepted unless it also supports use through the CLI.

The other issue is the possible raising of the barrier to contribution - not everyone understands classes but most understand functions.

TL;DR - Suggest holding off on refactoring around classes because this is potentially (likely?) a large undertaking with little payoff. Instead use tests to enforce/maintain consistency as the case may be.

**Comment by willu47** [comment]
Hi @ConnectedSystems - that is sound reasoning.  Would you like to add something to the 'development' section of the docs along this in your next PR?

**Comment by ConnectedSystems** [comment]
Sure thing.

Please note that in my travels, I found that some sampling methods are ignoring the seed argument when used via the CLI - fixing these as I go along.

**Comment by willu47** [comment]
Great, thanks @ConnectedSystems.

**Comment by ConnectedSystems** [comment]
Hi @willu47 

Close to finishing things off for this - I'm writing up some tests and will go back to add docstrings. 

I just want to double check something as I'm not a mathematician:
Are confidence intervals for the sensitivity indices expected to return the same values between runs if a seed value is being passed in? My gut feeling is that it should but it's not what I'm seeing.
(I've checked that the seed value is being passed in and handled correctly)

```bash
python ../SAlib/scripts/salib.py sample saltelli -p ../SALib/test_functions/params/Ishigami.txt -r model_input.txt -n 1000 --delimiter " " --precision 8 --max-order 2 --seed 100

python ../SAlib/scripts/salib.py analyze sobol -p ../SALib/test_functions/params/Ishigami.txt -Y model_output.txt -c 0 --max-order 2 -r 1000
```

First run:
```
Parameter S1 S1_conf ST ST_conf
x1 0.307975 0.064849 0.560137 0.085929
x2 0.447767 0.053084 0.438722 0.039276
x3 -0.004255 0.060072 0.242845 0.026576

Parameter_1 Parameter_2 S2 S2_conf
x1 x2 0.012205 0.083681
x1 x3 0.251526 0.107293
x2 x3 -0.009954 0.068971
```

Second run:
```
Parameter S1 S1_conf ST ST_conf
x1 0.307975 0.062046 0.560137 0.087442
x2 0.447767 0.053143 0.438722 0.041945
x3 -0.004255 0.058998 0.242845 0.025800

Parameter_1 Parameter_2 S2 S2_conf
x1 x2 0.012205 0.085130
x1 x3 0.251526 0.106961
x2 x3 -0.009954 0.066553
```

**Comment by jdherman** [comment]
Hi @ConnectedSystems , thanks again for your thoughtful work on this. I haven't followed the full discussion, but I can comment on this specific issue..

Yes you're right, in theory we would like the CIs to be repeatable with the same random seed. And this will happen if the sample/analyze steps are run from the same .py script.

But in this case from the command line, the interpreter is invoked twice, and the seed is only set for the sampling step. Nothing is preserving the RNG state for the analyze step. I think the only way around this would be to allow the analyze() function (for Sobol and any other method that computes a CI) to accept a random seed as an input.

I'd leave it to you as to whether that is worth implementing. I would hope that with enough bootstrap resamples, the differences in the CIs between random seeds would be relatively small.

By the way, in the examples folder we have a bunch of command-line examples in the .sh scripts. As part of the CLI update would you be able to update those as well? (https://github.com/SALib/SALib/tree/master/examples)

Thanks again!!

**Comment by ConnectedSystems** [comment]
Hi again,

Just wanted to let you know that in getting the CLI to work as expected I ended up edited the `setup.py` file directly; it seemed the `setup.cfg` file was being ignored. I realise now that pyscaffold is being used to generate the setup file (and future use would overwrite my changes). 

I don't have any experience with pyscaffold so I'm not sure how to set things up for it to generate what I've included. If I have time I will look into it over the next week.
